mobile_cv/torch/utils_pytorch/shareables.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import logging
+import os
+import pickle
+from multiprocessing import shared_memory
+from typing import Any, List, Tuple, Union
+
+import mobile_cv.torch.utils_pytorch.comm as comm
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+class _SharedMemoryRef(object):
+    """Deal with the clean up of shared memory"""
+
+    def __init__(self, shm: shared_memory.SharedMemory, owner_pid: int):
+        self.shm = shm
+        self.owner_pid = owner_pid
+
+    def __del__(self):
```

Comment:
Python does not guarantee to always call `__del__` upon exit - would that cause memory leak then?

There is a `SharedMemoryManager` in python that uses `ResourceTracker` - which starts a separate process whose only job is to reliably free the shm.. I haven't used it but may be relevant.

Reply:
By memory leak do you mean when data loader goes out of context, the memory won't be freed? (eg. during the training we run multiple evaluations, and each eval creates a test loader, then the )

Reply:
@ppwwyyxx I found that the `SharedMemoryManager` is a bit slow to start (it starts a new process) from GPU worker process whereas `SharedMemory` is instant, so I ended up not using it. Wondering what makes `SharedMemoryManager` more reliable for freeing shm, does it use refcount such that when all reference of shm goes out of context, the memory will be freed? I'm under the impression that `SharedMemory` can do the same, if calling `x, _ = share_numpy_array_locally(x)` with current implementation plus removing the `__del__`, when trying to access `x`, it'll raise segmentation fault because the shm has gone out-of-context and refcount is 0.

Reply:
The main problem with refcount is that it relies on executing Python code. When Python exits ungracefully, refcount doesn't work and `__del__` won't be called. `ResourceTracker` starts a separate Python process so it won't be affected by crashes in other Python process.

Took another look at `SharedMemory`. It seems when you create a `SharedMemory`, it already uses `ResourceTracker`: https://github.com/python/cpython/blob/c0859743d9ad3bbd4c021200f4162cfeadc0c17a/Lib/multiprocessing/shared_memory.py#L120 , which will start a new process. So there is probably no leak.

However, this should also be slow to start then.

Reply:
That makes sense. I ran some more benchmarks (multiple evaluations during training which creates test loader), there seems no memory leak.